src/ahp/prover.rs

```diff
@@ -457,29 +440,79 @@ impl<F: PrimeField> AHPForR1CS<F> {
                 *a *= b;
                 *a -= c * d;
             });
-        let rhs = r_alpha_evals.interpolate();
-        let q_1 = mask_poly.polynomial() + &rhs;
-        end_timer!(q_1_time);
+        let outer_poly = r_alpha_evals.interpolate_by_ref();
+        end_timer!(outer_poly_time);
+
+        let z_1_time = start_timer!(|| "Compute z_1 poly");
+
+        // In order to re-use the outer_poly_evals for computing z_1 (because z_1 shall be
+        // computed over domain H, while outer_poly_evals are on mul_domain), we ensure that
+        // H is a sub-domain of mul_domain.
+        assert_eq!(mul_domain.size() % domain_h.size(), 0);
+        let step = mul_domain.size()/domain_h.size();
+        assert_eq!(domain_h.group_gen(), mul_domain.group_gen().pow(&[step as u64]));
+
+        let outer_poly_evals_on_H = EvaluationsOnDomain::from_vec_and_domain(
+            r_alpha_evals.evals.clone().into_iter().step_by(step).collect(),
+            domain_h.clone()
+        );
+        let mut z_1 = BoundaryPolynomial::from_coboundary_polynomial_evals(
+            outer_poly_evals_on_H
+        ).unwrap().polynomial();
+
+        // Let b = zk_bound. This blinding polynomial is of type
+        // (c_0 + ... c_(2b-1) * X ^ (2b - 1) )
+        // since the boundary poly is queried twice as much as the others
+        let mut randomization_poly = Polynomial::rand(2 * zk_bound - 1, rng);
+        randomization_poly = randomization_poly.mul_by_vanishing_poly(domain_h.size());
+
+        // Add the blinding polynomial to z_1
+        z_1 = &z_1 + &randomization_poly;
+        end_timer!(z_1_time);
+
+        assert!(z_1.degree() <= domain_h.size() + 2 * zk_bound - 1);
+
+        let z_1_g_time = start_timer!(|| "Compute z_1_g poly");
+        let z_1_g = {
+
+            // z_1 has higher degree with respect to |H| due to randomization;
+            // therefore, when constructing z_1_g we have to take care of the
+            // higher powers of g. So the g vector will be:
+            // (1, g, g^2,...., g^n-1, 1, g, ..., g^((deg(z1) - 1) - |H|)
+            let size_diff = z_1.coeffs.len() as i32 - domain_h.size() as i32;
 
-        let sumcheck_time = start_timer!(|| "Compute sumcheck h and g polys");
-        let (h_1, x_g_1) = q_1.divide_by_vanishing_poly(&domain_h).unwrap();
-        let g_1 = Polynomial::from_coefficients_slice(&x_g_1.coeffs[1..]);
-        end_timer!(sumcheck_time);
+            // Exclude extraordinary high zk_bound
+            assert!(size_diff >= 0 && size_diff <= domain_h.size() as i32);
+
+            let mut g_s = domain_h.elements().collect::<Vec<F>>();
+            g_s.append(&mut g_s[..size_diff as usize].to_vec());
+
+            let mut z_1_g = z_1.clone();
+            z_1_g.coeffs.par_iter_mut().zip(g_s).for_each(|(a, b)| { *a *= b } );
```

Comment:
Possibly vector operation. How big the amount of coeffs?

Reply:
As for the other cases. Altough this one is an elementwise field mul of two vecs of size 2^19-2^20, it is probably not worth to move that to GPU.